chain release2.9 compatibility  Grace Period and overdue

0oM4R opened the pull request:
### Description

We have new accurate way to resume/unlock contracts on tfchain2.9, this pull request is for replace the old way as it is no longer supported.

#### General idea:
Now the amount required to resume the contract is described as **Overdue**.
The overdue amount basically is the sum of three parts:
  1. Total over draft: is the sum of additional overdraft and standard overdraft.
  2. Unbilled NU: is the unbilled amount of network usage.
  3. The estimated cost of the contract for the total period: this part is dependent on the contract type and if the contract is on a rented node or not.
> Note:
> - The total period is the time since the last billing added to the allowance period.
> - If the contract is a rent contract, will add both of ipv4 cost and the total overdue of all associated contracts. 



### Changes
**Please node**: you may notice that we have multiple functions that generally do the same thing but require different arguments. 
for example: we need some information that is already in the contract type, 
so instead of requesting the contract in side the function we are passing it.
but in some cases we only have the contract id? we provide a function that take the contract id and then request the contract then pass it to the inner function.
 
#### TFChain client
1. Fixed the [generateTypes.bash](https://github.com/threefoldtech/tfgrid-sdk-ts/pull/3387/files#diff-b4638b04c8f8163527467a64e7acb4b40072b9697f55b635c1672fbd2015c4f7):
  - the old script required to have ts-node installed, now we can run those commands using yarn.
2. In contracts, we add some functions that help in grid client new features.
#### Grid client: 
  - add deprecated annotation to the no longer supported functions that are associated with locked amount.
  - Calculator module:  added a function to calculate the name pricing.
  - Contract module: 
     - added some functions that calculate overdue, you can pass the contract, or the contract id.
     - add functions to unlock contracts, all contracts, and batch unlock using array of contracts or array of contractss ids
    - we are using gridproxy in unlock contracts
 
#### Dashboard

support new functions in grid client, refactor, and remove the deprecated gird client functions.
we have some changes in the flow:
 - If the user select node contracts, and some of them are on a rent contract, we will show the rent contract id with the selected contracts
     ![Screenshot from 2024-09-09 15-18-12](https://github.com/user-attachments/assets/355d2228-f041-4b18-82e8-0c1ae23e8bb7)

 - if use click on grace period chip of node contract with ipv4 on rented node the content of the dialog depends on the rent contract state:
    - if the rent contract  state is `Created` : will get the node contract overdue, and user will be able to unlock the contract.
        
    - If on `Grace period`: will show a message that you will have to check the rent contract and unlock it.
        ![Screenshot from 2024-09-10 15-06-56](https://github.com/user-attachments/assets/451b4149-bfed-46fe-ba04-8e0a9fdc0a9e)

   
fix loading spinner in loading locked amount details: 
 ![Screenshot from 2024-09-09 14-55-38](https://github.com/user-attachments/assets/6df51220-4b1e-4e68-ada8-4f55835338ea)



### Related Issues

#3340 
#3364 


### Tested Scenarios
- show contract/s cost : this will be nearly from pricing calculator if the contract got billed right before moving to the grace period as in this case overdue will be small for example, name contract cost should be about `0.025` tfts
- select node contracts on rented node that its contract on grace period,
- unlock all contracts associated with rent contract: in this case to verify that the node contract got billed, use `smarContractModule/contractPaymentState` call, the additional and standred overdraft should be 0
- Unlock single or multiple contracts
- in any case we should'n have a contract got billed twice: you can add log line to log the ids in batchUnlock contracts
Tips: 
- To send contract to a grace period, you need to follow some steps 
   1- empty you walet, this can be done by polkadot UI `Balances/TransferAll`
   2- get the contract Id, must be billable one like rent contract or node contract on shared node, name contract, and call `smarContractsModule/billcontractForBlock` using any other funded account.
   
- to send node contract with public ip on rented node to grace period and keep it's rent contract is on created state
   - fund you wallet, and call bill contract for block for the rent contract,
   - then send the node contract with ipv4 only to grace period as mentioned
 
### Documentation PR

For UI changes, Please provide the Documetation PR on [info_grid](https://github.com/threefoldtech/info_grid)




### Checklist

- [ ] Tests included 
- [x] Build pass
- [ ] Documentation
- [x] Code format and docstrings
- [x] Screenshots/Video attached (needed for UI changes)

0oM4R committed:
Refactor: add yarn to script to avoid errors

0oM4R committed:
Feat: add some methods and interfaces
- a method to get NU billing information
- a method to get contract payment state
- a method to get used resources by the node contract
- some related interfaces

0oM4R committed:
Feat: calculate unique name prcice

0oM4R committed:
chore: inhance docstirng

0oM4R committed:
Feat: support contract payment state calculations

0oM4R committed:
Refactor: unlock contracts to use new overdue calculations

0oM4R committed:
Refactor:
enhance code readablity.
reduce number of requests by passing the contract info to all chiled functions

0oM4R committed:
chore: inhance imports

0oM4R committed:
Refactor: unlock contracts
- use proxy url in unlockMycontracts
- refactor unlockContracts to accept array of contracts
- add function to unlock contracts by their ids

0oM4R committed:
add overdue details interfaces

0oM4R committed:
chore: contract overdue
- add function that takes only contract to unlock it
- add calculate overdue functions

0oM4R committed:
chore: expose get over due amount by the contract info only

0oM4R committed:
fix: remove duplicated call

0oM4R committed:
refactor:
mUSD converter will return the price in mTFT
fix unique name price calc
fix imports

0oM4R committed:
WIP: support overdue changes on contracts list in dashboard

0oM4R committed:
refactor: fix unique name contract

0oM4R committed:
wip fix calculations

0oM4R committed:
feat: support Payment state
- add contracts on rented node to the rent node overdue

0oM4R committed:
chore: clean up

0oM4R committed:
refactor: enhance contract lock state
it is storing the rent contract on a map by the node id, and ignore the unwanted calculations

0oM4R committed:
refactor: enhance contract lock
unlock dialog now will check if the node contract on rented node or not; if so will retrive the rent contract cost and store the and will avoid duplication

0oM4R committed:
refactor: fix is on rented node flag

0oM4R committed:
refactor: call the associated rent contract on unlock rent contracts

0oM4R committed:
feat: show the rent contract that will be unlocked as it is associated with selected contracts

0oM4R committed:
Style: fix loading spinner in lock dialogs

0oM4R committed:
fix: build

0oM4R committed:
fix: passing contracts to client

0oM4R committed:
docs: add deprecated annotation

0oM4R committed:
return the contract cost whatever if it is on a rented node or not

0oM4R committed:
docs: WIP adding docstrings

0oM4R committed:
WIP: support unlock node contract if the associated rent contract is in created state

0oM4R committed:
docs: WIP adding docstrings

0oM4R committed:
cleanup: remove unused module

0oM4R committed:
fix: build
remove unused import

0oM4R committed:
Chore: support unlock node contract if its rent contract is in created state

0oM4R committed:
Chore: avoid bill same contract multiple times

AhmedHanafy725 commented on packages/grid_client/src/clients/tf-grid/contracts.ts:
```diff
@@ -301,13 +331,216 @@ class TFContracts extends Contracts {
+    } catch (error) {
+      throw new GridClientError(`Failed to convert to mTFT due: ${error}`);
+    }
+  }
```
please use the currency module instead

AhmedHanafy725 commented on packages/grid_client/src/clients/tf-grid/contracts.ts:
```diff
@@ -301,13 +331,216 @@ class TFContracts extends Contracts {
+   * the IPV4 cost to add to the estimated cost of the rent contract.
+   */
+  private async getContractsCostOnRentedNode(nodeId: number, proxy: GridProxyClient) {
+    const contracts = await proxy.contracts.list({ nodeId, state: [ContractState.GracePeriod], numberOfPublicIps: 1 });
```
I think this only returns the first page(50 items), what if I have more than that?

AhmedHanafy725 commented on packages/grid_client/src/clients/tf-grid/contracts.ts:
```diff
@@ -301,13 +331,216 @@ class TFContracts extends Contracts {
+   * the function will convert them to GB and calculate the cost
+   * @returns { Object } contains the `dedicatedPrice` and the `sharedPrice` both are numbers
+   */
+  private async calcResourcesPrice(options: CalcResourcesPrice) {
+    const { cru, hru, sru, mru, calculator, ipv4, certified } = options;
+
+    const mruInGB = mru / Math.pow(1024, 3);
+    const sruInGB = sru / Math.pow(1024, 3);
+    const hruInGB = hru / Math.pow(1024, 3);
+
+    const res = await calculator.calculateWithMyBalance({
+      cru,
+      mru: mruInGB,
+      sru: sruInGB,
+      hru: hruInGB,
+      certified,
+      ipv4u: !!ipv4,
+    });
+
+    return {
+      dedicatedPrice: res.dedicatedPrice,
+      sharedPrice: res.sharedPrice,
+    };
+  }
```
Why do you add a calculate helper to the contracts module?

0oM4R replied:
this is a private function to reduce redundant code and make the conversion from bytes to GB inside it instead of doing it multiple times, should be moved to the calculator module ?

AhmedHanafy725 replied:
yes, it would be a better place

AhmedHanafy725 commented on packages/grid_client/src/clients/tf-grid/contracts.ts:
```diff
@@ -335,28 +568,60 @@ class TFContracts extends Contracts {
+
+    const contracts = await proxy.contracts.list({
+      state: [ContractState.GracePeriod],
+      twinId: await this.client.twins.getMyTwinId(),
```
this will be only one page

0oM4R committed:
Chore: use currency module to convert usd to tft

0oM4R committed:
refactor:
use the calculator module,
remove the added pricing related interfaces and functions

add helper method to convert from bytes to gb

0oM4R committed:
chore: list all contracts
to avoid default page size 50, add a function to work around that.

0oM4R committed:
fix: build
return if the contract list is empty

sameh-farouk commented on packages/grid_client/src/clients/tf-grid/contracts.ts:
```diff
@@ -301,13 +326,216 @@ class TFContracts extends Contracts {
+    return {
+      //return ip cost per month
+      totalIpCost: totalIPCost * 24 * 30,
+      totalOverdraft: totalOverDraft.add(totalNUCost),
```
You can't add them directly because `totalNUCost` is in unit-USD and `totalOverDraft` is in unit-TFT.

sameh-farouk commented on packages/grid_client/src/clients/tf-grid/contracts.ts:
```diff
@@ -301,13 +326,216 @@ class TFContracts extends Contracts {
+      new Decimal(0),
+    );
+    const totalNUCost = billingInfoResult.reduce((acc: number, billingInfo) => acc + billingInfo.amountUnbilled, 0);
+    const totalIPCost = (ipPrice * (contracts.data.length || 0)) / 1e7;
```
`ipPrice` was previously converted from unit-USD to USD. Why divide it by 1e7 again here?

sameh-farouk commented on packages/grid_client/src/clients/tf-grid/contracts.ts:
```diff
@@ -301,13 +326,216 @@ class TFContracts extends Contracts {
+   * @description
+   *  Name contract cost is fixed price for the unique name,
+   * Rent contract cost is the cost of the total node resources,
+   * Node contract have to cases:
```
There is a small typo here to -> two

sameh-farouk commented on packages/grid_client/src/clients/tf-grid/contracts.ts:
```diff
@@ -301,13 +326,216 @@ class TFContracts extends Contracts {
+      new Decimal(0),
+    );
+    const totalNUCost = billingInfoResult.reduce((acc: number, billingInfo) => acc + billingInfo.amountUnbilled, 0);
+    const totalIPCost = (ipPrice * (contracts.data.length || 0)) / 1e7;
```
By multiplying by `contracts.data.length`, are we assuming that a node contract can only have 1 IP address?

0oM4R replied:
the current behavior, yes, each contract should have 1 IP

0oM4R replied:
now I get the ip counts not the contracts

sameh-farouk commented on packages/grid_client/src/clients/tf-grid/contracts.ts:
```diff
@@ -301,13 +326,216 @@ class TFContracts extends Contracts {
+    /** Other contract types need the node information */
+    const nodeDetails = await proxy.nodes.byId(contract.details.nodeId);
+
+    const certified = nodeDetails.certificationType == CertificationType.Certified ? true : false;
```
Consider using the name `isCertified` instead.

sameh-farouk commented on packages/grid_client/src/clients/tf-grid/contracts.ts:
```diff
@@ -301,13 +326,216 @@ class TFContracts extends Contracts {
+   * please note that the unbilled NU amount is added to the total overdraft exactly as the in the other contract types.
+   * the IPV4 cost to add to the estimated cost of the rent contract.
+   */
+  private async getContractsCostOnRentedNode(nodeId: number, proxy: GridProxyClient) {
```
This function calculates two separate costs:
1. The overdue amount for all node contracts associated with a rent contract (total overdraft + unbilled NU cost).
2. The IP cost for all node contracts per month.

In my opinion, it would be better to handle these calculations with separate functions.

0oM4R replied:
Actually, I put both of them in one place to reduce listing the contracts again, here already have the contracts and their count, i was thinking about move the ip calculations to separate function and call it here but this will change nothing imo, what do you think?

sameh-farouk replied:
This is because this function does a lot of things.
What about splitting it into three separate functions?
The three new functions would be:
1. `getNodeContractsOnRentedNode`
2. `calculateNodeContractsOverdue`
3. `calculateIPCostPerMonth`
This way, you can retrieve the contracts from the first function and then pass them to the second and third functions.

sameh-farouk reviewed (commented):
I submitted a partial review today and will continue tomorrow.

sameh-farouk commented on packages/grid_client/src/clients/tf-grid/contracts.ts:
```diff
@@ -301,13 +326,216 @@ class TFContracts extends Contracts {
+      state: [ContractState.GracePeriod],
+      numberOfPublicIps: 1,
+    });
+
```
Can we handle the case where there are no returned contracts and return early?

sameh-farouk commented on packages/grid_client/src/clients/tf-grid/contracts.ts:
```diff
@@ -301,13 +326,216 @@ class TFContracts extends Contracts {
+      if (!contract.details.number_of_public_ips) return 0;
+
+      const ipPrice = (await this.client.pricingPolicies.get({ id: 1 })).ipu.value / 10 ** 7;
+      return ipPrice * 30 * 24;
```
Why not use the number of public IPs in the contract instead?
FYI, from the tfchain side, I can create a contract that reserves more than one public IP. Since it is a valid case, despite some client limitations or restrictions, it should be handled when calculating the contract costs as it may be created with different clients.

sameh-farouk commented on packages/grid_client/src/clients/tf-grid/contracts.ts:
```diff
@@ -301,13 +326,216 @@ class TFContracts extends Contracts {
+    /** Other contract types need the node information */
+    const nodeDetails = await proxy.nodes.byId(contract.details.nodeId);
+
+    const certified = nodeDetails.certificationType == CertificationType.Certified ? true : false;
```
Do you need the ternary operator here?
The condition `nodeDetails.certificationType == CertificationType.Certified` already evaluates to true or false.

Also, it is better to use === to avoid unexpected behavior caused by implicit type conversion.

sameh-farouk commented on packages/grid_client/src/clients/tf-grid/contracts.ts:
```diff
@@ -301,13 +326,216 @@ class TFContracts extends Contracts {
+
+    // cost of the current billing period with the mentioned allowance time
+    const totalPeriodCost = contractCostTFT.times(totalPeriodTime);
+    const overdue = totalOverDraft.add(unbilledNU);
```
As previously explained in my other comment, You can't add them directly.

sameh-farouk replied:
Another point to consider regarding unbilled NU is that it may be eligible for a discount or subject to an added premium:
* A discount may apply based on the user's balance.
* A premium is added when a certified node is used.

This has already been handled for overdraft stored on the contract payment state, as well as the cost of the contract retrieved from the calc module.

sameh-farouk commented on packages/grid_client/src/clients/tf-grid/contracts.ts:
```diff
@@ -301,13 +326,216 @@ class TFContracts extends Contracts {
+    const { standardOverdraft, additionalOverdraft, lastUpdatedSeconds } =
+      await this.client.contracts.getContractPaymentState(contractInfo.contract_id);
+
+    /**Calculate the elapsed seconds since last pilling*/
```
There is a small typo here "pilling" -> "billing"

sameh-farouk commented on packages/grid_client/src/clients/tf-grid/contracts.ts:
```diff
@@ -301,13 +326,216 @@ class TFContracts extends Contracts {
+    const overdue = totalOverDraft.add(unbilledNU);
+
+    //convert to TFT
+    const OverdueTFT = overdue.div(10 ** 7);
```
`OverdueTFT` name should follow camelCase convention.

0oM4R committed:
Chore: apply comments
- fix typos
- fix calculation in ipv4 cost
- fix nu cost calaculations

sameh-farouk commented on packages/grid_client/src/clients/tf-grid/contracts.ts:
```diff
@@ -301,13 +326,216 @@ class TFContracts extends Contracts {
+    if (nodeDetails.rented) {
+      if (!contract.details.number_of_public_ips) return 0;
+
+      const ipPrice = (await this.client.pricingPolicies.get({ id: 1 })).ipu.value / 10 ** 7;
```
I would avoid using magic numbers like 10 ** 7 and define it as constant `TFT_CONVERSION_FACTOR` (maybe in the client) then use it everywhere else for clarity and maintainability.

0oM4R replied:
add for the current file, will open an issue to make it global value for the grid client

0oM4R committed:
Chore: support multiple ipv4 per contract

0oM4R committed:
Chore: enahnce code readability
separate node contracts on rented node cost calculations to multiple functions
create constants with numbers repeatedly  used in the calculations
enhance docstring

0oM4R committed:
Chore:
fix typo
fix build
fix name convention

0oM4R committed:
Fix: avoid having deleted rent contracts while listing rent contracts

0oM4R committed:
Fix: pass CRU as number without conversion

0oM4R committed:
Chore: add node extra fees to rent contract cost

0oM4R committed:
refactor: include premuim price for certified nodes in the unbuiled nu

sameh-farouk commented:
@0oM4R I attempted a quick test, but it didn't go well.

![Screenshot_20240930_160919](https://github.com/user-attachments/assets/c5f9b2dc-5d04-40d9-a95e-c8c20b59c96a)

![Screenshot_20240930_160829](https://github.com/user-attachments/assets/e91264b9-9142-4508-a716-67016bf48224)

sameh-farouk commented:
Another point, it seems you haven't incorporated the changes discussed in the tg group yet.

> Unlock All/ Unlock -> Resume All / Resume
Locked Balance -> On hold
Show in UI: Available and On hold

0oM4R committed:
Merge branch 'development' of github.com:threefoldtech/tfgrid-sdk-ts into development_chain_release2.9

0oM4R commented:
> Another point, it seems you haven't incorporated the changes discussed in the tg group yet.
> 
> > Unlock All/ Unlock -> Resume All / Resume
> > Locked Balance -> On hold
> > Show in UI: Available and On hold

there is a separate issue for it https://github.com/threefoldtech/tfgrid-sdk-ts/issues/3400

0oM4R commented:
![Screenshot from 2024-09-30 17-59-09](https://github.com/user-attachments/assets/2df3ca54-78db-4e4d-9be8-3ac38edd1b67)
![Screenshot from 2024-09-30 17-59-42](https://github.com/user-attachments/assets/83775a37-91e2-4190-8253-beca30b19edf)

works fine with me but we should handle the zero tft case in the ui


> @0oM4R I attempted a quick test, but it didn't go well.
> 
> ![Screenshot_20240930_160919](https://private-user-images.githubusercontent.com/42457449/372117884-c5f9b2dc-5d04-40d9-a95e-c8c20b59c96a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc3MDYwMjgsIm5iZiI6MTcyNzcwNTcyOCwicGF0aCI6Ii80MjQ1NzQ0OS8zNzIxMTc4ODQtYzVmOWIyZGMtNWQwNC00MGQ5LWE5NWUtYzhjMjBiNTljOTZhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTMwVDE0MTUyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWUyNDEyZDkxZDIwMDRlMmY5N2Y2OGUxZDdhNTU5NjJlMzIzMzZkYmM1MzUyN2Q4MTYyNWJiZmJmNDgzMmVmNjMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.1e4jzbpnsWLmup6IzzgCzplxw1dSSp_QxehaFZ6CESw)
> 
> ![Screenshot_20240930_160829](https://private-user-images.githubusercontent.com/42457449/372117985-e91264b9-9142-4508-a716-67016bf48224.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc3MDYwMjgsIm5iZiI6MTcyNzcwNTcyOCwicGF0aCI6Ii80MjQ1NzQ0OS8zNzIxMTc5ODUtZTkxMjY0YjktOTE0Mi00NTA4LWE3MTYtNjcwMTZiZjQ4MjI0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTMwVDE0MTUyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTExZTc2MGFkN2VmMzI0N2ZhZDVmYzA2NTI1OTU5N2MwMzJhMzU0YjIyNWE2YzcxMDNkZmMxODY3ODNlMzZjNzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.JdlBKQ_5D_PyGEJ_gj5CqnHj9BnJ6G2Z-9c54CXvGPk)

sameh-farouk commented:
Sorry, My bad.
I didn't pull the latest changes. So I will give it another try tomorrow.

sameh-farouk commented:
> works fine with me but we should handle the zero tft case in the ui

Sorry, My bad.
I didn't pull the latest changes. So I will give it another try tomorrow.

0oM4R committed:
chore: set decimals to 7 in convert tft price

sameh-farouk commented:
Calculations looks much better now.

0oM4R committed:
refactor: convert the monthly cost to avoid missing decimals on converting cost per second

0oM4R commented:
Test report: 
will update this comment with the tested scenarios in details
 
#### Rent contract without any node contracts on it
![image](https://github.com/user-attachments/assets/f1b60e58-3e85-4c0d-a566-3b785d34193f)
![image](https://github.com/user-attachments/assets/33b389fa-bc2c-40f1-99b7-aaa22c380256)

### the allowance window in the next scenarios is __60__ seconds

0oM4R committed:
fix: get total overdue
skip the node contracts on rented nodes as it already calculated in rent contract

0oM4R committed:
refactor: calculate node contract on rented node
we were ignoring the elapsed seconds for node contracts and we assume that it is the same as rent contract

0oM4R committed:
refactor
reorder the calculation logic to be easier in debugging

0oM4R committed:
refactor:
pass the contracts by id as the contracts do not have the public_ips_number field

0oM4R committed:
fix: reset total overdue on refresh

0oM4R committed:
reset rentcontract on reset table

0oM4R committed:
Fix: add premium price for ip price on rent node

0oM4R committed:
Chore: remove unused functions

0oM4R committed:
Chore: set to fixed point 15 and fix the error msg

0oM4R committed:
Chore: include node contract on rented node in the getTotalOverDue function this will be needed for UI

0oM4R committed:
Style: rephrase unlock dialog msg

0oM4R committed:
merge development
